idlparser/src/types/mod.rs

```diff
@@ -0,0 +1,374 @@
+            "i8" => Some(PrimitiveType::I8),
+            "i16" => Some(PrimitiveType::I16),
+            "i32" => Some(PrimitiveType::I32),
+            "i64" => Some(PrimitiveType::I64),
+            "i128" => Some(PrimitiveType::I128),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum TypeDef {
+    Struct(StructDef),
+    Enum(EnumDef),
+}
+
+#[derive(Debug, PartialEq)]
+pub struct StructDef {
+    fields: Vec<StructField>,
+}
+
+impl StructDef {
+    pub(crate) fn new(fields: Vec<StructField>) -> Self {
+        Self { fields }
+    }
+
+    pub fn fields(&self) -> &[StructField] {
+        &self.fields
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct StructField {
+    name: Option<String>,
+    r#type: TypeDecl,
+}
+
+impl StructField {
+    pub(crate) fn new(name: Option<String>, r#type: TypeDecl) -> Self {
+        Self { name, r#type }
+    }
+
+    pub fn name(&self) -> Option<&str> {
+        self.name.as_deref()
+    }
+
+    pub fn r#type(&self) -> &TypeDecl {
+        &self.r#type
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct EnumDef {
+    variants: Vec<EnumVariant>,
+}
+
+impl EnumDef {
+    pub(crate) fn new(variants: Vec<EnumVariant>) -> Self {
+        Self { variants }
+    }
+
+    pub fn variants(&self) -> &[EnumVariant] {
+        &self.variants
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct EnumVariant {
+    name: String,
+    r#type: Option<TypeDecl>,
+}
+
+impl EnumVariant {
+    pub(crate) fn new(name: String, r#type: Option<TypeDecl>) -> Self {
+        Self { name, r#type }
+    }
+
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    pub fn r#type(&self) -> Option<&TypeDecl> {
+        self.r#type.as_ref()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn parser_works() {
+        let program_idl = r"
+          type ThisThatSvcAppTupleStruct = struct {
+            bool,
+          };
+          
+          type ThisThatSvcAppDoThatParam = struct {
+            p1: u32,
+            p2: str,
+            p3: ThisThatSvcAppManyVariants,
+          };
+          
+          type ThisThatSvcAppManyVariants = enum {
+            One,
+            Two: u32,
+            Three: opt u32,
+            Four: struct { a: u32, b: opt u16 },
+            Five: struct { str, u32 },
+            Six: struct { u32 },
+          };
+  
+          service {
+            DoThis : (p1: u32, p2: str, p3: struct { opt str, u8 }, p4: ThisThatSvcAppTupleStruct) -> struct { str, u32 },
+            DoThat : (param: ThisThatSvcAppDoThatParam) -> result (struct { str, u32 }, struct { str }),
+            query This : (v1: vec u16) -> u32,
+            query That : (v1: null) -> result (str, str),
+          };
+  
+          type T = enum { One }
+        ";
+
+        let program = parse_idl(program_idl).unwrap();
+
+        assert_eq!(program.types().len(), 4);
+
+        //println!("ast: {:#?}", program);
```

Comment:
```suggestion
```

Reply:
I've used https://crates.io/crates/insta in my PR just to keep track of this output changing. You probably want this: https://docs.rs/insta/latest/insta/macro.assert_debug_snapshot.html

Reply:
This is not for tracking changes in output, but for seeing what it looks like